### PR TITLE
*: support get-old-kv in watch

### PR DESCRIPTION
**Pull request by xiang90**
/cc @heyitsanthony 

I will rename the preserve-kv in delete to get-old-kv too. also i will probably add the same option for put.

### CODE REVIEW

**Comment by xiang90** [comment]
@ajityagaty You might want to review this.

**Comment by ajityagaty** [comment]
@xiang90 LGTM
This addresses #4620 right?

**Comment by chancez** [comment]
Isn't `old` is a bit vague? Would `prev` or `previous` be more descriptive perhaps?

File: mvcc/mvccpb/kv.proto [hunk 1]
```diff
@@ -43,4 +43,7 @@ message Event {
   // A DELETE/EXPIRE event contains the deleted key with
   // its modification revision set to the revision of deletion.
   KeyValue kv = 2;
+
+  // oldkv holds the KeyValue before the event happens when requested.
+  KeyValue oldKV = 3;
```
Comment (heyitsanthony) [inline mvcc/mvccpb/kv.proto]: `prev_kv`?

File: etcdserver/etcdserverpb/rpc.proto [hunk 2]
```diff
@@ -574,6 +574,10 @@ message WatchCreateRequest {
   }
   // filters filter the events at server side before it sends back to the watcher.
   repeated FilterType filters = 5;
+
+  // If getOldKV is set, created watcher gets the old KV before the event happens.
+  // If the old KV is already compacted, nothing will be returned.
+  bool getOldKV = 6;
```
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: just `prev_kv`? the `get` bit seems unnecessary-- `WithPrevKV()` seems more natural than `WithGetOldKV()`

**Comment by xiang90** [comment]
@chancez @heyitsanthony OK. I think we can rename old_kv to prev_kv. For delete, the original idea was named it as `deleted_kv` to be more specific and clear. But that does not make sense any more since we decided to provide the prev_kv for watch/delete/put.

File: etcdserver/api/v3rpc/watch.go [hunk 3]
```diff
@@ -253,6 +268,14 @@ func (sws *serverWatchStream) sendLoop() {
 			events := make([]*mvccpb.Event, len(evs))
 			for i := range evs {
 				events[i] = &evs[i]
+
+				if sws.getOldKV[wresp.WatchID] {
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/watch.go]: do map lookup before loop?

**Comment by heyitsanthony** [comment]
approach looks fine, but not a fan of the naming

**Comment by xiang90** [comment]
@heyitsanthony @chancez All fixed. PTAL.

File: etcdserver/etcdserverpb/rpc.proto [hunk 4]
```diff
@@ -574,6 +574,10 @@ message WatchCreateRequest {
   }
   // filters filter the events at server side before it sends back to the watcher.
   repeated FilterType filters = 5;
+
+  // If prevKV is set, created watcher gets the previous KV before the event happens.
+  // If the previous KV is already compacted, nothing will be returned.
+  bool prevKV = 6;
```
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: should this be `prev_kv`, not `prevKV`?

File: mvcc/mvccpb/kv.proto [hunk 5]
```diff
@@ -43,4 +43,7 @@ message Event {
   // A DELETE/EXPIRE event contains the deleted key with
   // its modification revision set to the revision of deletion.
   KeyValue kv = 2;
+
+  // prevKV holds the key-value pair before the event happens.
+  KeyValue prevKV = 3;
```
Comment (heyitsanthony) [inline mvcc/mvccpb/kv.proto]: s/prevKV/prev_kv/ ?

**Comment by heyitsanthony** [comment]
lgtm after protobuf naming style fixups

File: etcdserver/etcdserverpb/rpc.proto (hunk 4 above)
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: sure.

**Pull request merged**